### PR TITLE
[megatron, fsdp, doc] feat: implement GPG loss. Add GPG advantage estimator implementation.

### PR DESCRIPTION
**Pull request by diqiuzhuanzhuan**
…and integrate into PPO training scripts and core algorithms

### Checklist Before Starting

- [x] Searched for similar PR(s).
- [x] Checked PR Title format
  - In format of: [modules] type: Title
  - modules are in `fsdp, megatron, sglang, vllm, rollout, trainer, ci, training_utils, recipe, hardware, deployment, ray, worker, single_controller, misc, perf, model, algo, env, tool, ckpt, doc, data`
  - type is in `feat, fix, refactor, chore`
  - can involve multiple modules, seperated by `,` or space, like `[megatron, fsdp, doc] feat: xxx`

### What does this PR do?
Implement GPG loss (GPG: A Simple and Strong Reinforcement Learning Baseline for Model Reasoning) which can achieve comparable performance in less training time.

### Test
some training records:
image
image

### Specific Changes

> List the specific changes.
Add doc of GPG in docs/algo/gpg.md
Add the addvantage estimation function of gpg in verl/trainer/ppo/core_algos.py.
Add compute_gpg_loss function of gpg in verl/ trainer/ppo/core_algos.py.
Add a conditional branch to determine whether to use the GPG loss in verl/workers/actor/dp_actor.py and megatron_actor.py
Add example scripts of GPG in examples/gpg_trainer. 

### Usage Example

```shell
# Add code snippet or script demonstrating how to use this 
bash examples/gpg_trainer/run_qwen2-7b_math.sh
```

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [x] Add `[BREAKING]` to the PR title `description` if it breaks any API.
- [x] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).

### CODE REVIEW

**Review by tongyx361 — APPROVED** [review]

File: docs/algo/gpg.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# GPG: A Simple and Strong Reinforcement Learning Baseline for Model Reasoning
```
Comment (eric-haibin-lin) [inline docs/algo/gpg.md]: pls include it in index.rst

Comment (eric-haibin-lin) [reply docs/algo/gpg.md]: please make a copy of this file in examples/gpg_trainer as well, thx

File: verl/workers/actor/megatron_actor.py [hunk 2]
```diff
@@ -356,17 +357,35 @@ def loss_func(output, data, meta_info):
...
-                    log_prob=log_prob,
-                    advantages=advantages,
-                    response_mask=response_mask,
-                    cliprange=clip_ratio,
-                    cliprange_low=clip_ratio_low,
-                    cliprange_high=clip_ratio_high,
-                    clip_ratio_c=clip_ratio_c,
-                    loss_agg_mode=loss_agg_mode,
-                )
+
```
Comment (eric-haibin-lin) [inline verl/workers/actor/megatron_actor.py]: could you add the reference training performance and log in https://github.com/volcengine/verl/blob/main/docs/algo/baseline.md ? Since both megatron and FSDP actors are changed, please include both in `baseline.md`. 
In general we do not expect a large number of feature/algorithms supported in megatron due to the maintainance overhead. It's totally fine if you only add this to fsdp actor only. However once the feature is added, we'll ask for reference training logs

**Review by eric-haibin-lin — CHANGES_REQUESTED** [review]

File: docs/algo/gpg.md (hunk 1 above)
Comment (diqiuzhuanzhuan) [reply docs/algo/gpg.md]: Ok, now all the required tasks are completed. Please take a look.

File: verl/workers/actor/megatron_actor.py (hunk 2 above)
Comment (diqiuzhuanzhuan) [reply verl/workers/actor/megatron_actor.py:360]: Hi, all updates are done. Both Megatron and FSDP logs have been added to baseline.md.
Let me know if further changes are needed. Thanks!

Comment (diqiuzhuanzhuan) [reply verl/workers/actor/megatron_actor.py]: I dismissed the stale review after addressing all comments and pushing new commits.
Please feel free to take another look when convenient. Thanks!

File: verl/trainer/config/ppo_trainer.yaml [hunk 3]
```diff
@@ -210,6 +210,9 @@ actor_rollout_ref:
     # Whether to use KL loss instead of KL reward penalty. True for GRPO
```
Comment (eric-haibin-lin) [inline verl/trainer/config/ppo_trainer.yaml]: what about the following:
- reuse `policy_loss. loss_mode == 'gpg'` instead of adding `use_gpg_loss`
- change the call `policy_loss_fn(old_log_prob, log_prob, advantages, response_mask, loss_agg_mode, self.config)` to include all keywords as well (old_log_prob=old_log_prob, log_prob=log_prob, ...)
- compute_gpg_loss still returns 4 tensors, with the latter 3 being torch nan tensor if they're not meaningful
- keep `verl/workers/actor/dp_actor.py` the same as verl main

Comment (diqiuzhuanzhuan) [reply verl/trainer/config/ppo_trainer.yaml]: Sure, I was thinking the same thing.

Comment (diqiuzhuanzhuan) [reply verl/trainer/config/ppo_trainer.yaml]: Hi, I’ve addressed all the requested changes. Please take another look.

**Review by eric-haibin-lin — APPROVED** [review]

**Comment by diqiuzhuanzhuan** [comment]
Hi, this PR includes a GitHub Actions workflow and is awaiting approval. Could a maintainer please review and approve the workflow run? Thanks!!

**Comment by eric-haibin-lin** [comment]
The failed test looks irrelevant. Well done and thanks for the contribution!

**Comment by eric-haibin-lin** [comment]
We're moving all example scripts from fsdp1 to fsdp2. Would you mind updating the example script and the record in docs/algo/baseline.md with strategy=fsdp2? Thanks! @diqiuzhuanzhuan

**Comment by diqiuzhuanzhuan** [comment]
> We're moving all example scripts from fsdp1 to fsdp2. Would you mind updating the example script and the record in docs/algo/baseline.md with strategy=fsdp2? Thanks! @diqiuzhuanzhuan
Sure, I’ll take care of that. Thanks for the heads-up! I’ll update the example script and the record in docs/algo/baseline.md with strategy=fsdp2.
